kernel/LinearAlgebra/SmallVector.h

```diff
@@ -36,137 +36,190 @@
  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
-#ifndef HPGEM_KERNEL_SMALLVECTOR_H
-#define HPGEM_KERNEL_SMALLVECTOR_H
+#ifndef HPGEM_KERNEL_GSmallVector_H
+#define HPGEM_KERNEL_GSmallVector_H
+
+// Note: there is an depenency between MiddleSizeVector and SmallVector to allow
+// conversion.
+#include "MiddleSizeVector.h"
 
 #include "Logger.h"
+#include <algorithm>
 #include <array>
 #include <cmath>
-#include "MiddleSizeVector.h"
-#include <algorithm>
+#include <functional>
 #include <numeric>
 
 namespace hpgem {
 namespace LinearAlgebra {
-class MiddleSizeVector;
-/// \class SmallVector
-/// \brief This is a vector of doubles
+
+namespace Detail {
+// Several helper functions for GSmallVector.
+//   - CrossProduct requires partial specialization
+//
+template <typename T>
+void computeCrossProduct(const GSmallVector<2, T>&, const GSmallVector<2, T>&,
+                         GSmallVector<2, T>&);
+template <typename T>
+void computeCrossProduct(const GSmallVector<3, T>&, const GSmallVector<3, T>&,
+                         GSmallVector<3, T>&);
+template <typename std::size_t nRows, typename T>
+void computeCrossProduct(const GSmallVector<nRows, T>&,
+                         const GSmallVector<nRows, T>&,
+                         GSmallVector<nRows, T>&);
+
+inline MiddleSizeVector::type convertToMiddleSize(std::complex<double> v);
+inline MiddleSizeVector::type convertToMiddleSize(double v);
+// As out argument to allow overloading
+inline void convertFromMiddleSize(MiddleSizeVector::type v, double& out);
+inline void convertFromMiddleSize(MiddleSizeVector::type v,
+                                  std::complex<double>& out);
+
+inline double conj(double v) { return v; }
+inline std::complex<double> conj(std::complex<double> v) {
+    return std::conj(v);
+}
+
+inline std::size_t hash(double v) noexcept { return std::hash<double>()(v); }
+inline std::size_t hash(std::complex<double> v) noexcept {
+
+    return std::hash<double>()(v.real()) + 17 * std::hash<double>()(v.imag());
```

Comment:
why 17?

Reply:
Just any other prime number to mix the hash values, no further particular reason.